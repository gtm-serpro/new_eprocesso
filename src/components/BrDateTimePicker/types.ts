export type DTPickerType = "date" | "time" | "dateTime";
export type DTPickerInput = string | number | Date;
export type DTPickerOutput = Date[] | string;

export type DateOptions = {
  minDate?: string | Date;
  maxDate?: string | Date;
};

export type TimeOptions = {
  minuteIncrement?: number;
  defaultHour?: number;
  defaultMinute?: number;
};
