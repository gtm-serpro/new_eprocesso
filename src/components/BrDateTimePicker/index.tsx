import { useSpreadProps } from "@/hooks/useSpreadProps";
import { BRDateTimePicker } from "@govbr-ds/core";
import classNames from "classnames";
import monthSelectPlugin from "flatpickr/dist/plugins/monthSelect";
import "flatpickr/dist/plugins/monthSelect/style.css";
import React, { useEffect, useRef } from "react";
import { useMtProps } from "../../hooks/useMtProps";
import useUniqueId from "../../hooks/useUniqueId";
import IMtProps from "../../types/IMtProps";

export interface BrDateTimePickerProps
  extends Omit<React.InputHTMLAttributes<HTMLInputElement>, "onChange">,
    IMtProps {
  /**
   * Tipo do DateTimePicker:
   * - date: seleciona data (padrão).
   * - time: seleciona apenas hora.
   * - date-time: seleciona data e hora.
   * - month: seleciona mês e ano.
   */
  dataType?: "date" | "time" | "date-time" | "month";
  /**
   * Modo do Datetime.
   * - single: um valor apenas a escolher (padrão).
   * - range: um intervalo de valores a escolher.
   */
  dataMode?: "single" | "range";
  /**
   * Label do DatetimePicker.
   */
  label?: React.ReactNode;
  /**
   * Data mínima selecionável.
   */
  minDate?: string | Date;
  /**
   * Data máxima selecionável.
   */
  maxDate?: string | Date;
  /**
   * Incremento dos minutos.
   */
  minuteIncrement?: number;
  /**
   * Hora padrão no DateTimePicker.
   */
  defaultHour?: number;
  /**
   * Minuto padrão no DateTimePicker.
   */
  defaultMinute?: number;
  /**
   * Evento que emite o valor selecionado no DateTimePicker no tipo string.
   *
   * @param selectedDate - valor selecionado no DateTimePicker.
   */
  onChange?: (selectedDate: string) => void;
  /**
   * Evento que emite o valor selecionado no DateTimePicker no tipo Date.
   *
   * @param selectedDate - valor selecionado no DateTimePicker.
   */
  onDateChange?: (selectedDate: Date) => void;
  /**
   * Evento que emite o(s) valor(es) selecionado(s) no DateTimePicker no tipo Array de Date.
   * Útil quando o `dataMode` for "range".
   *
   * @param selectedDates - valores selecionados no DateTimePicker.
   */
  onRangeChange?: (selectedDates: Array<Date>) => void;
}

const dataTypeConverter = (dataType: BrDateTimePickerProps["dataType"]) => {
  switch (dataType) {
    case "time":
      return "time";
    case "date-time":
      return "datetime-local";
    default:
      return "text";
  }
};

export interface DateTimePickerRef extends HTMLInputElement {
  element: HTMLInputElement;
  wrapper: HTMLElement;
  inputWrapper: HTMLElement;
  label: HTMLElement;
  button: HTMLButtonElement;
}

const BrDateTimePicker = React.forwardRef<DateTimePickerRef, BrDateTimePickerProps>(
  (
    {
      className,
      id,
      children,
      dataMode = "single",
      dataType = "date",
      label,
      minDate,
      maxDate,
      minuteIncrement,
      defaultHour,
      defaultMinute,
      onChange,
      onDateChange,
      onRangeChange,
      ...props
    },
    ref
  ) => {
    const fid = useUniqueId(id, "datetimepicker_____");
    const refWrapper = useRef(null);
    const refObject = useRef<any>(null);

    const buttonIcon = dataType === "time" ? "clock" : "calendar-alt";

    if (dataType === "time" && dataMode === "range") {
      console.warn('O BrDateTimePicker com dataType="time" não suporta o dataMode="range".');
    }

    let placeholder = "";

    switch (dataType) {
      case "date":
        placeholder = dataMode === "range" ? "DD/MM/AAAA até DD/MM/AAAA" : "DD/MM/AAAA";
        break;
      case "time":
        placeholder = "HH:MM";
        break;
      case "date-time":
        placeholder =
          dataMode === "range" ? "DD/MM/AAAA HH:mm até DD/MM/AAAA HH:mm" : "DD/MM/AAAA HH:mm";
        break;
      case "month":
        placeholder = dataMode === "range" ? "Mês/Ano até Mês/Ano" : "Mês/Ano";
        break;
    }

    const handleInputChange = (e: React.ChangeEvent<HTMLInputElement>) => {
      const length = e.target.value.length;

      if (length === 0 || length === e.target.maxLength) {
        onChange?.(e.target.value);
        onDateChange?.(new Date(e.target.value));
        onRangeChange?.([new Date(e.target.value)]);
      }
    };

    // Inicializando o datetimepicker
    useEffect(() => {
      refObject.current = new BRDateTimePicker("br-datetimepicker", refWrapper.current, {
        ...(minDate && { minDate }),
        ...(maxDate && { maxDate }),
        ...(minuteIncrement && { minuteIncrement }),
        ...(defaultHour && { defaultHour }),
        ...(defaultMinute && { defaultMinute }),
        ...(dataType === "month" && { plugins: [monthSelectPlugin({ dateFormat: "F/Y" })] }),
        onChange: (dates: Date[], dateString: string) => {
          onRangeChange?.(dates);
          onDateChange?.(dates[0]);
          onChange?.(dateString);
        },
      });

      return () => refObject.current?.calendar?.destroy();
    }, [minDate, maxDate, minuteIncrement, defaultHour, defaultMinute, dataType, dataMode]);

    const mtProps = useMtProps(props);
    const spreadProps = useSpreadProps(props);

    return (
      <div
        className={classNames("br-datetimepicker", ...mtProps, className)}
        ref={refWrapper}
        data-mode={dataMode}
        data-type={dataTypeConverter(dataType)}
      >
        <div className="br-input has-icon">
          {label && <label htmlFor={fid}>{label}</label>}
          <input
            autoComplete="off"
            id={fid}
            type={dataTypeConverter(dataType)}
            placeholder={placeholder}
            data-input="data-input"
            data-toggle="data-toggle"
            {...spreadProps}
            onChange={handleInputChange}
          />
          <button
            className="br-button circle small"
            type="button"
            aria-label="Abrir Calendário"
            data-toggle="data-toggle"
            id="timepicker-input-btn"
          >
            <i className={`fas fa-${buttonIcon}`} aria-hidden="true"></i>
          </button>
          {children}
        </div>
      </div>
    );
  }
);

export default BrDateTimePicker;
