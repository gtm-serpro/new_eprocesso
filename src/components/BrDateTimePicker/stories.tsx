import { Meta, Story } from "@storybook/react";

import { Col } from "@/common/Grid";

import BrDateTimePicker, { BrDateTimePickerProps } from ".";

export default {
  title: "Formul√°rios/DateTimePicker",
  component: BrDateTimePicker,
} as Meta;

const Template: Story<BrDateTimePickerProps> = (args) => {
  return (
    <Col sm={6} md={4} lg={3}>
      <BrDateTimePicker {...args} />
    </Col>
  );
};

export const DatePicker = Template.bind({});
DatePicker.args = {
  dataType: "date",
  dataMode: "single",
  onChange: (value) => console.log("onChange", value),
  onDateChange: (value) => console.log("onDateChange", value),
  onRangeChange: (value) => console.log("onRangeChange", value),
};

const today = new Date();
const todayPlus15 = new Date();
todayPlus15.setDate(todayPlus15.getDate() + 15);

export const DatePickerRange = Template.bind({});
DatePickerRange.args = {
  ...DatePicker.args,
  dataType: "date",
  dataMode: "range",
  minDate: today,
  maxDate: todayPlus15,
};

export const TimePicker = Template.bind({});
TimePicker.args = {
  ...DatePicker.args,
  dataType: "time",
  dataMode: "single",
};

export const DateTimePicker = Template.bind({});
DateTimePicker.args = {
  ...DatePicker.args,
  dataType: "date-time",
  dataMode: "single",
};

export const MonthPicker = Template.bind({});
MonthPicker.args = {
  ...DatePicker.args,
  dataType: "month",
  dataMode: "single",
};
