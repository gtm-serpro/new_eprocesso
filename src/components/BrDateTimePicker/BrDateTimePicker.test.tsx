import { render, screen } from "@testing-library/react";

import BrDateTimePicker from ".";

describe("<BrDatePicker />", () => {
  it("deveria renderizar o date picker", () => {
    const { container } = render(<BrDateTimePicker onChange={() => ""} />);
    expect(container.firstChild).toHaveClass("br-datetimepicker");
    expect(container.firstChild).toMatchSnapshot();
  });

  it("deveria adicionar label informado", () => {
    const { container } = render(<BrDateTimePicker label="Meu BrDatePicker" onChange={() => ""} />);
    screen.getByRole("textbox", { name: "Meu BrDatePicker" });
    expect(container.firstChild).toMatchSnapshot();
  });
});
