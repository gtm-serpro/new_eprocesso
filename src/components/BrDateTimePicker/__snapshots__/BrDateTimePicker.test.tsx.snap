// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<BrDatePicker /> deveria adicionar label informado 1`] = `
<div
  class="br-datetimepicker"
  data-mode="single"
  data-type="text"
>
  <div
    class="br-input has-icon"
  >
    <label
      for="datetimepicker_____2"
    >
      Meu BrDatePicker
    </label>
    <input
      autocomplete="off"
      class="flatpickr-input"
      data-input="data-input"
      data-toggle="data-toggle"
      id="datetimepicker_____2"
      maxlength="10"
      placeholder="DD/MM/AAAA"
      type="text"
    />
    <button
      aria-hidden="true"
      aria-label="Abrir Calendário"
      class="br-button circle small"
      data-toggle="data-toggle"
      id="timepicker-input-btn"
      tab-index="-1"
      type="button"
    >
      <i
        aria-hidden="true"
        class="fas fa-calendar-alt"
      />
    </button>
  </div>
</div>
`;

exports[`<BrDatePicker /> deveria renderizar o date picker 1`] = `
<div
  class="br-datetimepicker"
  data-mode="single"
  data-type="text"
>
  <div
    class="br-input has-icon"
  >
    <input
      autocomplete="off"
      class="flatpickr-input"
      data-input="data-input"
      data-toggle="data-toggle"
      id="datetimepicker_____1"
      maxlength="10"
      placeholder="DD/MM/AAAA"
      type="text"
    />
    <button
      aria-hidden="true"
      aria-label="Abrir Calendário"
      class="br-button circle small"
      data-toggle="data-toggle"
      id="timepicker-input-btn"
      tab-index="-1"
      type="button"
    >
      <i
        aria-hidden="true"
        class="fas fa-calendar-alt"
      />
    </button>
  </div>
</div>
`;
